README.md
pyapns
======

A universal Apple Push Notification Service (APNS) provider.

Features:

 * **REST interface** so you can start sending notifications with any language immediately
 * Native Python API included
 * Scalable, fast and easy to distribute behind a proxy
 * Multi-application and dual environment support
 * Disconnection log interface which provides reasons for recent connection issues
 * Supports the full suite of APN gateway functionality

### Quick Start

Install and start the daemon:

    $ sudo easy_install pyapns
    $ wget https://raw.github.com/samuraisam/pyapns/master/pyapns.tac
    $ twistd -ny pyapns.tac 

Provision an app and send a notification:

    $ curl -d '{"certificate":"/path/to/certificate.pem"}'      \
          http://localhost:8088/apps/com.example.app/sandbox

    $ curl -d '{"token": "le_token",                            \
                "payload": {"aps": {"alert": "Hello!"}},        \
                "identifier": "xxx", "expiry": 0}'              \
          http://localhost:8088/apps/com.example.app/sandbox/notifications

### About

pyapns is a daemon that is installed on a server and designed to take the pain out of sending push notifications to Apple devices. Typically your applications you will have a thread that maintains an SSL socket to Apple. This can be error-prone, hard to maintain and plainly a burden your app servers should not have to deal with.

Additionally, pyapns provides several features you just wouldn't get with other solutions such as the disconnection log which remembers which notifications and tokens caused disconnections with Apple - thus allowing your application layer to make decisions about whether or not th continue sending those types of notifications. This also works great as a debugging layer.

pyapns supports sending notifications to multiple applications each with multiple environments. This is handy so you don't have to push around your APN certificates, just keep them all local to your pyapns installation.

## The REST interface

The rest interface is by default hosted on port `8088` and provides all of the functionality you'll need. Here are some basic things you need to know about how it works:

 * All functionality is underneath the `/apps` top level path
 * Functionality specific to individual apps is available underneath the `/apps/{app id}/{environment}` path where `app id` is the provisioned app id and `environment` is the Apple environment (either "sandbox" or "production")
 * You can get a list of provisioned apps: `GET /apps`
 * Objects all have a `type` attribute that tells you which kind of object it is
 * Successful responses will have a top level object with a `response` and `code` keys
 * Unsuccessful responses will have a top level object with an `error` and `code` keys

### Provisioning An App

Before sending notifications to devices, you must first upload your certificate file to the server so pyapns can successfully make a connection to the APN gateway. The certificates must be a PEM encoded file. [This](http://stackoverflow.com/questions/1762555/creating-pem-file-for-apns) stackoverflow answer contains an easy way to accomplish that.

You can upload the PEM and provision the app multiple ways:

 1. Send the PEM file directly when provisioning the apps. Just read the whole PEM file into memory and include it as the `certificate` key:
     
         $ curl -d '{"certificate": "$(cat /path/to/cert.pem)"}' $HOST:$PORT/apps/com.example.myid/production         

 2. Upload the PEM file ahead of time to the same server as the pyapns daemon and provide the path to the certificate as the `certificate` key:

         $ curl -d '{"certificate": "/path/to/cert.pem"}' $HOST:$PORT/apps/com.example.myid/production         

Notice above that we are including in the URL the app id desired as well as the environment desired. They are the last and 2nd-to-last elements of the path, respectively. So for this url the app id is _com.example.myid_ and the environment is _production_. Any time you access functionality specific to these apps you'll be accessing it as a subpath of this full path.

<dl>
<dt> GET /apps</dt>
<dd>
Returns a list of all provisioned apps

<strong>Example Response</strong>
```json
{
    "response": [
        {
            "type": "app",
            "certificate": "/path/to/cert.pem",
            "timeout": 15,
            "app_id": "my.app.id",
            "environment": "sandbox"
        }
    ],
    "code": 200
}
```
</dd>
</dl>
#### GET _/apps/:app_id/environment_

Returns information about a provisioned app

##### Example Response
```json
{
    "response": {
        "type": "app",
        "certificate": "/path/to/cert.pem",
        "timeout": 15,
        "app_id": "my.app.id",
        "environment": "sandbox"
    },
    "code": 200
}
```

#### POST _/apps/:app_id/:environment_

Creates a newly provisioned app. You can POST multiple times to the same URL and it will merely re-provision the app, taking into account the new certificate and timeout. There may be more config values to provision in the future.

###### Example Body:
```json
{
    "certificate": "certificate or path to certificate",
    "timeout":     15
}
```
##### Example Response
```json
{
    "response": {
        "type": "app",
        "certificate": "/path/to/cert.pem",
        "timeout": 15,
        "app_id": "my.app.id",
        "environment": "sandbox"
    },
    "code": 201
}
```

### Sending Notifications
###### Identifiers and Expiry

#### Retrieving Feedback

#### Retrieving Disconnection Events

### The Included Python API

### Installing in Production

To install in production, you will want a few things that aren't covered in the quickstart above:

 1. Automated provisioning of apps. This is supported when the pyapns server is started up.
 2. Install [python-epoll](http://pypi.python.org/pypi/python-epoll/) and [ujson](http://pypi.python.org/pypi/ujson) for dramatically improved performance
 3. (optional) start multiple instances behind a reverse proxy like HAProxy or Nginx

#### Automated provisioning

#### Production dependencies

#### Example `supervisord` config

#### Multiple instances behind a reverse proxy



CHANGES.txt
version 0.5.0 - 2013-01-25
==========================

There is lots of new stuff here but this release is backwards compatible. The
REST interface operates on a different default port and runs completey in 
paralell with the XML-RPC one. That said, REST is the way forward and all new
features will be added to that server.

New:

  * An all-new REST interface that simplifies access to pyapns's functionality.

    The XML-RPC server will be depreciated over the coming years. Next year it
    will be "depreciated" as in it will start giving off warnings and the year
    after it will likely be entirely removed.

  * REST interface uses Apple's new "enhanced" notification format which has 
    the ability to send back errors in the event Apple severs the connection 
    to pyapns (which can happen for a number of reasons).

  * A disconnection log which will pair disconnection events with the 
    notifications at fault. This will allow you to remove offending tokens
    and notifications.

  * The new Python client uses the very fine requests library which provides
    several advantages over the old XML-RPC library principally: connection
    keep-alive and reuse and connection pooling.

Fixed Bugs:

  * Fixed #31/#26 - Token blacklisting - we will not blacklist tokens as it's 
    not the job of this daemon to persist that kind of state. We however now 
    give you the tools to blacklist tokens should you wish to do so.

  * Fixed #14 - Apple Dropping Connection - this can happen for any number
    of reasons, by using the new REST service you can get an idea of why.

  * Fixed #10 - reprovisioning - while this issue isn't specifically to allow
    reprovisioning, the reason @mikeytrw wanted the feature was to replace
    certs, which is now possible.

  * Fixed #28 - sending multiple notifications after a disconnect fails - 
    simply fixed the way the deferred was handeled

version 0.4.0 - 2012-02-14
==========================

New:

  * Client interface methods have more explicit arguments and now take an 
    error callback.

    The `provision` `notify` and `feedback` methods in v0.3.2 would
    implicitly assume they should perform asynchronously if the `callback`
    argument was supplied. They would spin off a thread, later calling
    the callback.

    Now those methods take an `async` parameter, when True, will spin off
    the XML-RPC request in another thread, calling `callback(result)` when
    complete or `errback(exception)` if an error occurred.

  * Clients will now automatically try to re-provision themselves if `INITIAL`
    was provided during configuration. This helps, in case the server goes
    down, to be able to retry provisioning at least once.

Fixed bugs:

  * Fixed #21 - logging of full contents of file PEM file when providing the 
    whole thing in `provision()`

  * Fixed #25 - provide a wrapper for api functions which will automaticaly 
    try to reprovision should it encounter a `UknownAppID`.

Other:

  * Removed some of the redundant documentation from the setup.py file. 
    Eventually I want to move to sphynx docs. That software didn't really 
    exist in the same capacity in 2010 as it does now.

version 0.3.2 - 2010-08-05
==========================

Started tracking changes. This is the version that everyone has been using for
the past few years. 



pyapns/client.py
import xmlrpclib
import threading
import httplib
import functools
from sys import hexversion
import requests
from pyapns import _json as json
from pyapns.model import Notification, DisconnectionEvent


class ClientError(Exception):
    def __init__(self, message, response):
        super(ClientError, self).__init__(message)
        self.message = message
        self.response = response


class Client(object):
    @property
    def connection(self):
        con = getattr(self, '_connection', None)
        if con is None:
            con = self._connection = requests.Session()
        return con

    def __init__(self, host='http://localhost', port=8088, timeout=20):
        self.host = host.strip('/')
        self.port = port
        self.timeout = timeout

    def provision(self, app_id, environment, certificate, timeout=15):
        """
        Tells the pyapns server that we want set up an app to receive 
        notifications from us.

        :param app_id: An app id, you can use anything but it's
        recommended to just use the bundle identifier used in your app.
        :type app_id: string
        
        :param environment: Which environment are you using? This value
        must be either "production" or "sandbox".
        :type environment: string

        :param certificate: A path to a encoded, password-free .pem file 
        on the pyapns host. This must be a path local to the host! You 
        can also read an entire .pem file in and send it in this value 
        as well.
        :type certificate: string

        :returns: Dictionary-representation of the App record
        :rtype: dict
        """
        status, resp = self._request(
            'POST', 'apps/{}/{}'.format(app_id, environment),
            data={'certificate': certificate, 'timeout': timeout}
        )
        if status != 201:
            raise ClientError('Unable to provision app id', resp)
        return resp['response'] 

    def notify(self, app_id, environment, notifications):
        """
        Sends notifications to clients via the pyapns server. The 
        `app_id` and `environment` must be previously provisioned 
        values--either by using the :py:meth:`provision` method or 
        having been bootstrapped on the server.

        `notifications` is a list of notification dictionaries that all
        must have the following keys:

            *  `payload` is the actual notification dict to be jsonified
            *  `token` is the hexlified device token you scraped from 
                the client
            *  `identifier` is a unique id specific to this id. for this 
                you may use any value--pyapns will generate its own 
                internal ID to track it. The APN gateway only allows for
                this to be 4 bytes.
            *  `expiry` is how long the notification should be retried 
                for if for some reason the apple servers can not contact
                the device

        You can also construct a :py:class:`Notification` object--the
        dict and class representations are interchangable here.

        :param app_id: Which app id to use
        :type app_id: string

        :param environmenet: The environment for the app_id
        :type environment: string

        :param notifications: A list of notification dictionaries
        (see the discussion above)
        :type notifications: list

        :returns: Empty response--this method doesn't return anything
        :rtype: dict 
        """
        notes = []
        for note in notifications:
            if isinstance(note, dict):
                notes.append(Notification.from_simple(note))
            elif isinstance(note, Notification):
                notes.append(note)
            else:
                raise ValueError('Unknown notification: {}'.format(repr(note)))
        data = [n.to_simple() for n in notes]

        status, resp = self._request(
            'POST', 'apps/{}/{}/notifications'.format(app_id, environment), 
            data=data
        )
        if status != 201:
            raise ClientError('Could not send notifications', resp)
        return resp['response']

    def feedback(self, app_id, environment):
        """
        Gets the from the APN feedback service. These are tokens that 
        Apple considers to be "dead" - that you should no longer attempt
        to deliver to.

        Returns a list of dictionaries with the keys:

            * `timestamp` - the UTC timestamp when Apple determined the
               token to be dead
            * `token` - the hexlified version of the token

        :param app_id: Which app id to use
        :type app_id: string

        :param environmenet: The environment for the app_id
        :type environment: string

        :rtrype: list
        """
        status, feedbacks = self._request(
            'GET', 'apps/{}/{}/feedback'.format(app_id, environment)
        )
        if status != 200:
            raise ClientError('Could not fetch feedbacks', resp)
        return feedbacks['response']

    def disconnections(self, app_id, environment):
        """
        Retrieves a list of the 5000 most recent disconnection events 
        recorded by pyapns. Each time apple severs the connection with 
        pyapns it will try to send back an error packet describing which
        notification caused the error and the error that occurred.

        :param app_id: Which app id to use
        :type app_id: string

        :param environmenet: The environment for the app_id
        :type environment: string

        :rtype: list
        """
        status, disconnects = self._request(
            'GET', 'apps/{}/{}/disconnections'.format(app_id, environment)
        )
        if status != 200:
            raise ClientError('Could not retrieve disconnections')
        ret = []
        for evt in disconnects['response']:
            ret.append(DisconnectionEvent.from_simple(evt))
        return ret

    def _request(self, method, path, args=None, data=None):
        url = '{}:{}/{}'.format(self.host, self.port, path)
        kwargs = {'timeout': self.timeout}
        if args is not None:
            kwargs['params'] = args
        if data is not None:
            kwargs['data'] = json.dumps(data)

        func = getattr(self.connection, method.lower())
        resp = func(url, **kwargs)
        if resp.headers['content-type'].startswith('application/json'):
            resp_data = json.loads(resp.content)
        else:
            resp_data = None
        return resp.status_code, resp_data


## OLD XML-RPC INTERFACE ------------------------------------------------------

OPTIONS = {'CONFIGURED': False, 'TIMEOUT': 20}


def configure(opts):
    if not OPTIONS['CONFIGURED']:
        try:  # support for django
            import django.conf
            OPTIONS.update(django.conf.settings.PYAPNS_CONFIG)
            OPTIONS['CONFIGURED'] = True
        except:
            pass
        if not OPTIONS['CONFIGURED']:
            try:  # support for programatic configuration
                OPTIONS.update(opts)
                OPTIONS['CONFIGURED'] = True
            except:
                pass
        if not OPTIONS['CONFIGURED']:
            try:  # pylons support
                import pylons.config
                OPTIONS.update({'HOST': pylons.config.get('pyapns_host')})
                try:
                    OPTIONS.update(
                        {'TIMEOUT': int(pylons.config.get('pyapns_timeout'))})
                except:
                    pass  # ignore, an optional value
                OPTIONS['CONFIGURED'] = True
            except:
                pass
        # provision initial app_ids
        if 'INITIAL' in OPTIONS:
            for args in OPTIONS['INITIAL']:
                provision(*args)
    return OPTIONS['CONFIGURED']


class UnknownAppID(Exception):
    pass


class APNSNotConfigured(Exception):
    pass


def reprovision_and_retry(func):
    """
    Wraps the `errback` callback of the API functions, automatically trying to
    re-provision if the app ID can not be found during the operation. If that's
    unsuccessful, it will raise the UnknownAppID error.
    """
    @functools.wraps(func)
    def wrapper(*a, **kw):
        errback = kw.get('errback', None)
        if errback is None:
            def errback(e):
                raise e

        def errback_wrapper(e):
            if isinstance(e, UnknownAppID) and 'INITIAL' in OPTIONS:
                try:
                    for initial in OPTIONS['INITIAL']:
                        provision(
                            *initial)  # retry provisioning the initial setup
                    func(*a, **kw)  # and try the function once more
                except Exception, new_exc:
                    errback(new_exc)  # throwing the new exception
            else:
                errback(e)  # not an instance of UnknownAppID - nothing we can do here
        kw['errback'] = errback_wrapper
        return func(*a, **kw)
    return wrapper


def default_callback(func):
    @functools.wraps(func)
    def wrapper(*a, **kw):
        if 'callback' not in kw:
            kw['callback'] = lambda c: c
        return func(*a, **kw)
    return wrapper


@default_callback
@reprovision_and_retry
def provision(app_id, path_to_cert, environment, timeout=15, async=False,
              callback=None, errback=None):
    args = [app_id, path_to_cert, environment, timeout]
    f_args = ['provision', args, callback, errback]
    if not async:
        return _xmlrpc_thread(*f_args)
    t = threading.Thread(target=_xmlrpc_thread, args=f_args)
    t.daemon = True
    t.start()


@default_callback
@reprovision_and_retry
def notify(app_id, tokens, notifications, async=False, callback=None,
           errback=None):
    args = [app_id, tokens, notifications]
    f_args = ['notify', args, callback, errback]
    if not async:
        return _xmlrpc_thread(*f_args)
    t = threading.Thread(target=_xmlrpc_thread, args=f_args)
    t.daemon = True
    t.start()


@default_callback
@reprovision_and_retry
def feedback(app_id, async=False, callback=None, errback=None):
    args = [app_id]
    f_args = ['feedback', args, callback, errback]
    if not async:
        return _xmlrpc_thread(*f_args)
    t = threading.Thread(target=_xmlrpc_thread, args=f_args)
    t.daemon = True
    t.start()


def _xmlrpc_thread(method, args, callback, errback=None):
    if not configure({}):
        raise APNSNotConfigured('APNS Has not been configured.')
    proxy = ServerProxy(OPTIONS['HOST'], allow_none=True, use_datetime=True,
                        timeout=OPTIONS['TIMEOUT'])
    try:
        parts = method.strip().split('.')
        for part in parts:
            proxy = getattr(proxy, part)
        return callback(proxy(*args))
    except xmlrpclib.Fault, e:
        if e.faultCode == 404:
            e = UnknownAppID()
        if errback is not None:
            errback(e)
        else:
            raise e


## --------------------------------------------------------------
## Thank you Volodymyr Orlenko:
## http://blog.bjola.ca/2007/08/using-timeout-with-xmlrpclib.html
## --------------------------------------------------------------

def ServerProxy(url, *args, **kwargs):
    t = TimeoutTransport()
    t.timeout = kwargs.pop('timeout', 20)
    kwargs['transport'] = t
    return xmlrpclib.ServerProxy(url, *args, **kwargs)


class TimeoutTransport(xmlrpclib.Transport):
    def make_connection(self, host):
        if hexversion < 0x02070000:
            conn = TimeoutHTTP(host)
            conn.set_timeout(self.timeout)
        else:
            conn = TimeoutHTTPConnection(host)
            conn.timeout = self.timeout
        return conn


class TimeoutHTTPConnection(httplib.HTTPConnection):
    def connect(self):
        httplib.HTTPConnection.connect(self)
        self.sock.settimeout(self.timeout)


class TimeoutHTTP(httplib.HTTP):
    _connection_class = TimeoutHTTPConnection

    def set_timeout(self, timeout):
        self._conn.timeout = timeout
